src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1692 @@
+  ---------------------------------------------------------------*/
+
+int lsrkStep_ComputeNewDomEig(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem)
+{
+  int retval = SUN_SUCCESS;
+
+  if ((step_mem->is_ext_dom_eig))
+  {
+    retval = step_mem->extDomEig(ark_mem->tn, ark_mem->ycur, ark_mem->fn,
+                                 &step_mem->lambdaR, &step_mem->lambdaI,
+                                 ark_mem->user_data, ark_mem->tempv1,
+                                 ark_mem->tempv2, ark_mem->tempv3);
+    step_mem->num_dom_eig_updates++;
+    if (retval != ARK_SUCCESS)
+    {
+      arkProcessError(ark_mem, ARK_DOMEIG_FAIL, __LINE__, __func__, __FILE__,
+                      "Unable to estimate the dominant eigenvalue");
+      return (ARK_DOMEIG_FAIL);
+    }
+
+    if (step_mem->lambdaR * ark_mem->h > ZERO)
+    {
+      arkProcessError(NULL, ARK_DOMEIG_FAIL, __LINE__, __func__, __FILE__,
+                      "lambdaR*h must be nonpositive");
+      return (ARK_DOMEIG_FAIL);
+    }
+    else if (step_mem->lambdaR == 0 && SUNRabs(step_mem->lambdaI) > 0)
+    {
+      arkProcessError(NULL, ARK_DOMEIG_FAIL, __LINE__, __func__, __FILE__,
+                      "DomEig cannot be purely imaginary");
+      return (ARK_DOMEIG_FAIL);
+    }
+
+    step_mem->lambdaR *= step_mem->dom_eig_sfty;
+    step_mem->lambdaI *= step_mem->dom_eig_sfty;
+    step_mem->sprad =
+      SUNRsqrt(SUNSQR(step_mem->lambdaR) + SUNSQR(step_mem->lambdaI));
+  }
+  else
+  {
+    arkProcessError(NULL, ARK_DOMEIG_FAIL, __LINE__, __func__, __FILE__,
+                    "Internal dom_eig is not supported yet!");
+    return (ARK_DOMEIG_FAIL);
+  }
+  step_mem->dom_eig_is_current = SUNTRUE;
+
+  step_mem->spr_max = (step_mem->sprad > step_mem->spr_max) ? step_mem->sprad
+                                                            : step_mem->spr_max;
+
+  if (step_mem->sprad < step_mem->spr_min || ark_mem->nst == 0)
+  {
+    step_mem->spr_min = step_mem->sprad;
+  }
+
+  step_mem->new_dom_eig = SUNFALSE;
+
+  return retval;
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_DomEigUpdateLogic:
+
+  This routine checks if the step is accepted or not and reassigns
+  the dom_eig update flags accordingly.
+  ---------------------------------------------------------------*/
+
+void lsrkStep_DomEigUpdateLogic(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem,
+                                sunrealtype dsm)
+{
+  if (dsm <= ONE || ark_mem->fixedstep)
+  {
+    N_VScale(ONE, ark_mem->tempv2, ark_mem->fn);
+    ark_mem->fn_is_current = SUNTRUE;
+
+    step_mem->dom_eig_is_current = (step_mem->const_Jac == SUNTRUE);
+    step_mem->nstsig      = (step_mem->nstsig + 1) % step_mem->dom_eig_freq;
+    step_mem->new_dom_eig = SUNFALSE;
+    if (step_mem->nstsig == 0)
+    {
+      step_mem->new_dom_eig = !step_mem->dom_eig_is_current;
+    }
+  }
+  else { step_mem->new_dom_eig = !step_mem->dom_eig_is_current; }
```

Comment:
Ah, yes. I though this was an `else` to the first `if` above which it's not. 

If a step fails before this point (e.g., RHS recoverable return) or after (e.g., constraint handling) this is not accounted for. Should non-error test fails be considered? Dan's comment about moving this check to the start of a step would also make this easier to consider.

I just noticed that `dom_eig_freq` is used differently from how we use the "frequency" elsewhere. This updates every `dom_eig_freq` steps regardless of when the eigenvalue was last computed instead of `dom_eig_freq` steps from the last update. For example, the Jacobian update the check is `ark_mem->nst >= arkls_mem->nstlj + arkls_mem->msbj` where `nstlj` is the step number of the update and `msbj` is the frequency. I this this should use similar logic.

Reply:
Agreed.  I think that moving the check to the start of the step, and using a similar mechanism as we use for `nstlj`, would be the best strategy (particularly since @gardner48 correctly noted that the current implementation would "count" rejected steps due to constraint handling).

Reply:
`step_mem->nstsig` counter is updated to restart with a zero with each dominant eigenvalue computation. I think it should be using the exact same logic now.